proposals/migrate-over-ipfs.md

```diff
@@ -0,0 +1,163 @@
+- This would prove out IPFS as a way to ship code/updates to users.
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+This is ones step towards our goal of updating go-ipfs itself over IPFS. In terms of knowledge, not much.
+
+#### Confidence
+_How sure are we that this is the right problem to tackle? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+3?
+
+- It's a frequent source of upgrade problems: https://github.com/ipfs/ipfs-desktop/issues?q=is%3Aissue+fs-repo-migrations
+- It's preventing us from shipping new go-ipfs releases (with important fixes) to brave users.
+- It hurts adoption in China.
+
+## Project definition
+
+#### Brief plan of attack
+
+**Design sketch:**
+
+Given a config section like:
+
+```js
+{
+    "Migration": {
+        // Sources in order of preference where "HTTPS" means our gateways and "IPFS" means over IPFS.
+        // (an empty list means "do the default thing")
+        "DownloadPolicy": ["HTTPS", "custom-gateway.io", "IPFS"],
+        // Whether or not to keep the migration after downloading it.
+        "Keep": "pin"|"cache"|"discard"
+    }
+}
+```
+
+For the initial version, the default download policy should be `["HTTPS", "IPFS"]` unless the local
+node uses private networks (has a swarm key) in which case it should just be `["HTTPS"]` for safety.
+As we gain confidence in our ability to reliably download migrations over IPFS, we should make it
+the default.
+
+When migrating a repo, until we succeed.
+
+1. If the next source is HTTPS, download from `ipfs.io`.
+2. If the source is IPFS:
+    1. Create a new _temporary_ repo (go-ipfs can't read the current repo because it's an older version).
+    2. Start a new _temporary_ go-ipfs node in the temporary repo.
+        a. This node should not listen for inbound connections as it has no way to know which ports/transports should be configured (can't read the config).
+        b. This node should not expose an API/gateway.
+    3. Download the required migration binaries using the temporary go-ipfs node.
+4. Migrate the main go-ipfs node's repo.
+5. Start the main go-ipfs node.
+6. Maybe store a copy of the migration:
+    1. If `Keep` is not "discard", import the migration binary into the main go-ipfs node's repo.
+    2. If `Keep` is "pin", pin the migration binary.
+7. If running, shutdown any temporary nodes and delete the temporary repo.
+
+**Implementation steps:**
+
+1. Implement everything but step 6.
+2. Implement step 6. Unless nodes keep around a copy of the migration, this feature isn't going to be all that useful.
+3. Eventually, switch over to using IPFS by default instead of the gateway as the first option.
+
+NOTE: the only reason we're proposing using the gateway first is that the migration blocks startup so it needs to be fast and reliable.
```

Comment:
What if we set up peering to the relevant gateway or cluster nodes so we defaulted to IPFS and dogfood Bitswap? On the surface this seems easy, is it worthwhile?

Reply:
I'd like to _start_ with HTTPs by default so we can get this in ASAP. Then, maybe in the next release, switch to IPFS by default so we have some time to test.

But yeah, peering by default sounds like a great idea.

Reply:
@aschmahmann could you add that to the proposal?